### PR TITLE
build: use Task

### PR DESCRIPTION
**Pull request by fortuna**
We now build outline-ss-server directly and cross-compile the docker image.

### CODE REVIEW

**Review by sbruens — CHANGES_REQUESTED** [review]
Can you get rid of `scripts/run_action.sh` now?

File: src/metrics_server/README.md [hunk 1]
```diff
@@ -67,20 +67,20 @@ npm run action metrics_server/start
...
 
 - Unit test
   ```sh
-  npm run action metrics_server/test
+  task metrics_server:test
   ```
 - Integration test
   ```sh
-  npm run action metrics_server/test_integration
+  task metrics_server:integration:test
```
Comment (jyyi1) [inline src/metrics_server/README.md:85]: *nit* shall we rename it to `integration_test`? Cuz we don't have other sub-tasks for `integration`?

File: src/metrics_server/Taskfile.yml [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+    deps: [{task: build, vars: {BUILD_MODE: prod}}]
+    cmds:
+      - gcloud app deploy '{{.TASKFILE_DIR}}/dispatch.yaml' '{{joinPath .SERVER_DIR "prod"}}' --project uproxysite --verbosity info --no-promote --no-stop-previous-version
+
+  start:
+    desc: Start the metrics server locally
+    deps: [build]
+    cmds:
+      - cp '{{.TASKFILE_DIR}}/config_dev.json' '{{.SERVER_DIR}}/config.json'
+      - npx node '{{.SERVER_DIR}}/index.js'
```
Comment (jyyi1) [inline src/metrics_server/Taskfile.yml:42]: Do we need `npx` here? Can we just use `node {{.SERVER_DIR}}/index.js`?

File: src/shadowbox/Taskfile.yml [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+  BUILD_DIR: '{{joinPath .USER_WORKING_DIR "build" "shadowbox"}}'
+
+tasks:
+  build:
+    desc: Build the Outline Server Node.js app
+    vars:
+      TARGET_OS: '{{.TARGET_OS | default "linux"}}'
+      TARGET_ARCH: '{{.TARGET_ARCH | default "x86_64"}}'
+      GOOS: '{{get (dict "macos" "darwin") .TARGET_OS | default .TARGET_OS}}'
+      GOARCH: '{{get (dict "x86_64" "arm64") .TARGET_ARCH | default .TARGET_ARCH}}'
```
Comment (jyyi1) [inline src/shadowbox/Taskfile.yml:13]:
```suggestion
      GOARCH: '{{get (dict "x86_64" "amd64") .TARGET_ARCH | default .TARGET_ARCH}}'
```

**Review by jyyi1 — COMMENTED** [review]
This is cool. Maybe we can support building from Windows once we migrate to a cross-platform build pipeline in the future.

File: src/metrics_server/Taskfile.yml (hunk 2 above)
Comment (fortuna) [reply src/metrics_server/Taskfile.yml]: Removed

File: src/shadowbox/Taskfile.yml (hunk 3 above)
Comment (fortuna) [reply src/shadowbox/Taskfile.yml]: Fixed

File: src/shadowbox/Taskfile.yml [hunk 4]
```diff
@@ -0,0 +1,119 @@
...
+      - {task: build, vars: {TARGET_OS: linux, TARGET_ARCH: '{{.TARGET_ARCH}}', TARGET_DIR: '{{joinPath .IMAGE_ROOT "/opt/outline-server"}}'}}
+      - cp -R '{{joinPath .TASKFILE_DIR "scripts"}}' '{{.IMAGE_ROOT}}/scripts'
+      - mkdir -p '{{joinPath .IMAGE_ROOT "/etc/periodic/weekly"}}'
+      - cp '{{joinPath .TASKFILE_DIR "scripts" "update_mmdb.sh"}}' '{{joinPath .IMAGE_ROOT "/etc/periodic/weekly/"}}'
+      # Create default state directory
+      - mkdir -p '{{joinPath .IMAGE_ROOT "/root/shadowbox/persisted-state"}}'
+      # Copy entrypoint command
+      - cp '{{joinPath .TASKFILE_DIR "docker/cmd.sh"}}' '{{.IMAGE_ROOT}}/'
+      # Build image with given root
+      - podman build --force-rm
```
[the remark this answers is not in the document]
Comment (fortuna) [reply src/shadowbox/Taskfile.yml]: I agree. It was for testing, I didn't actually intend to change. I've reverted to docker, but have an option to override. I can now do `./task shadowbox:docker:build DOCKER=podman`

[the remark this answers is not in the document]
Comment (fortuna) [reply src/shadowbox/Taskfile.yml]: Ahh, I forgot about it. I'll work on it.

File: src/shadowbox/README.md [hunk 5]
```diff
@@ -43,13 +43,13 @@ The Outline Server, internal name "Shadowbox," is designed to streamline the set
...
      ```sh
-     npm run action shadowbox/server/start
+     task shadowbox:start
      ```
 
    - **Docker Container**
 
      ```sh
-     npm run action shadowbox/docker/start
+     task shadowbox:docker:start
```
[the remark this answers is not in the document]
Comment (fortuna) [reply src/shadowbox/README.md:52]: Ah, I forgot about it. I'll work on it.

File: src/shadowbox/Taskfile.yml (hunk 4 above)
[the remark this answers is not in the document]
Comment (sbruens) [reply src/shadowbox/Taskfile.yml]: Nice, the override is cool.

**Review by sbruens — APPROVED** [review]
LGTM but please don't forget to add the `docker:start` task.

File: Taskfile.yml [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+
+  shadowbox:
+    taskfile: ./src/shadowbox/Taskfile.yml
+    vars: {OUTPUT_BASE: '{{joinPath .BUILD_ROOT "shadowbox"}}'}
+
+tasks:
+  clean:
+    desc: Clean output files
+    cmds:
+      - rm -rf .task src/*/node_modules/ build/ node_modules/ third_party/shellcheck/download/ third_party/*/bin third_party/jsign/*.jar
```
Comment (jyyi1) [inline Taskfile.yml:29]: Do we need to remove the `./task` binary built by `npm install` as well?

**Review by jyyi1 — APPROVED** [review]
The code LGTM, please make sure the CI job passes.

File: Taskfile.yml (hunk 6 above)
Comment (fortuna) [reply Taskfile.yml]: Makes sense. Done

**Comment by fortuna** [comment]
docker:start added. It was a pain to make it work. Ran into cross-compilation issues...

**Pull request merged**